src/protocol/serial.rs

```diff
@@ -0,0 +1,307 @@
+                    .map(|option| {
+                        option.map(|(ptr, len)| unsafe { slice::from_raw_parts(ptr, len) })
+                    })
+                    .map_err(|e| {
+                        ::get_system_table().boot_services().free_pool(buf_ptr);
+                        e
+                    })
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+pub struct SerialIOProtocol {
+    raw_protocol: &'static RawSerialIOProtocol,
+    baud_rate: Option<u64>,
+    receive_fifo_depth: Option<u32>,
+    timeout: Option<u32>,
+    parity: Option<ParityType>,
+    data_bits: Option<u8>,
+    stop_bits: Option<StopBits>,
+}
+
+impl SerialIOProtocol {
+    /// Create and reset a new SerialIOProtocol struct.
+    pub fn new() -> Result<SerialIOProtocol, Status> {
+        let bs = ::get_system_table().boot_services();
+        bs.locate_protocol::<RawSerialIOProtocol>(0 as *const CVoid)
+            .and_then(|protocol| {
+                if let Err(e) = protocol.reset() {
+                    return Err(e);
+                }
+
+                Ok(SerialIOProtocol {
+                    raw_protocol: protocol,
+                    baud_rate: None,
+                    receive_fifo_depth: None,
+                    timeout: None,
+                    parity: None,
+                    data_bits: None,
+                    stop_bits: None,
+                })
+            })
+    }
+
+    fn set_attributes(&self) -> Result<(), Status> {
+        match self.raw_protocol.set_attributes(
+            self.baud_rate,
+            self.receive_fifo_depth,
+            self.timeout,
+            self.parity,
+            self.data_bits,
+            self.stop_bits,
+        ) {
+            Status::Success => Ok(()),
+            e => Err(e),
+        }
+    }
+
+    /// Update this serial device's attributes. Unlike the UEFI protocol, this function can be
+    /// called incrementally - that is, passing `None` for any attribute will leave it unchanged
+    /// from its current value, instead of resetting it to the device default.
+    pub fn update_attributes(
+        &mut self,
+        baud_rate: Option<u64>,
+        receive_fifo_depth: Option<u32>,
+        timeout: Option<u32>,
+        parity: Option<ParityType>,
+        data_bits: Option<u8>,
+        stop_bits: Option<StopBits>,
+    ) -> Result<(), Status> {
+        self.baud_rate = baud_rate.or(self.baud_rate);
+        self.receive_fifo_depth = receive_fifo_depth.or(self.receive_fifo_depth);
+        self.timeout = timeout.or(self.timeout);
+        self.parity = parity.or(self.parity);
+        self.data_bits = data_bits.or(self.data_bits);
+        self.stop_bits = stop_bits.or(self.stop_bits);
+
+        self.set_attributes()
+    }
+
+    /// Write a string to the serial device.
+    pub fn write(&self, data: &str) -> Result<usize, Status> {
+        // GRUB sets the attributes on the serial device on each read or write, so we will too.
```

Comment:
We're also doing it because it avoids a specific problem.  Please document what that is.